game/songs.cc

```diff
@@ -453,24 +481,34 @@ void Songs::typeCycle(unsigned short cat) {
 }
 
 std::string Songs::getSortDescription() const {
-	std::string str;
-	switch (m_order) {
-	  case 0: str = _("random order"); break;
-	  case 1: str = _("sorted by song"); break;
-	  case 2: str = _("sorted by artist"); break;
-	  case 3: str = _("sorted by edition"); break;
-	  case 4: str = _("sorted by genre"); break;
-	  case 5: str = _("sorted by path"); break;
-	  case 6: str = _("sorted by language"); break;
-	  case 7: str = _("sorted by score"); break;
-	  default: throw std::logic_error("Internal error: unknown sort order in Songs::getSortDescription");
-	}
-	return str;
+	if(m_order < m_songOrders.size())
+		return m_songOrders[m_order]->getDescription();
+
+	throw std::logic_error("Internal error: unknown sort order in Songs::getSortDescription");
 }
 
 void Songs::sortChange(Game& game, SortChange diff) {
-	m_order = static_cast<unsigned short>(m_order + to_underlying(diff)) % orders;
-	if (m_order >= orders) m_order += orders;
+	auto const orders = static_cast<decltype(m_order)>(m_songOrders.size());
+
+	switch(diff) {
+		case SortChange::BACK:
+			if(m_order == 0) {
+				m_order = orders - 1;
+			}
+			else {
+				--m_order;
+			}
+		break;
+		case SortChange::FORWARD:
+			if(++m_order == orders) {
+				m_order = 0;
+			}
+		break;
+		case SortChange::RESET:
+			m_order = 0;
+		break;
+	}
+
```

Comment:
This whole method can be simplified
```
void Songs::sortChange(Game& game, SortChange diff) {
    auto const numOrders = m_songOrders.size();

    switch (diff) {
        case SortChange::BACK:
            m_order = (m_order == 0) ? numOrders - 1 : m_order - 1;
            break;
        case SortChange::FORWARD:
            m_order = (m_order + 1) % numOrders;
            break;
        case SortChange::RESET:
            m_order = 0;
            break;
    }

    RestoreSel restore(*this);
    config["songs/sort-order"].ui() = m_order;

    if (m_order == 1 || m_order == 2 || m_order == 3 || m_order == 4 || m_order == 6) {
        UnicodeUtil::m_sortCollator->setStrength(config["game/case-sorting"].b() ? icu::Collator::TERTIARY : icu::Collator::SECONDARY);
    }

    sort_internal();
    writeConfig(game, false);
}

```
* Easier to read switch
* Changed second switch into an if statement
    * which can be even more simplified: `if(m_order >= 1 && m_order <=6)`

Reply:
* I wrote a class Cycle that replaces the unsigned short. Looks much better.
* The second switch is nonsense and must be moved to the order classes. Advantage: Now the sequence can be changed without changing the switch or if or ...